### ISSUE TITLE
DictionaryEquality is redundant

### ISSUE THREAD

**Issue by PatrikBak** [diegofrata/Generator.Equals#5]
We can juse use `UnorderedSequenceEquality`, since `IDictionary<TKey, TValue>` extends `IEnumerable<KeyValuePair<TKey, TValue>>`, and `KeyValuePair<TKey, TValue>` implements `Equals` and `HashCode` correctly. 

```CSharp
using Generator.Equals;
using System;
using System.Collections.Generic;
using System.Collections.Immutable;

namespace Test
{
    [Equatable]
    partial record MyRecord(
        [property: UnorderedSequenceEquality] Dictionary<string, int> Mutable,
        [property: UnorderedSequenceEquality] ImmutableDictionary<string, int> Immutable
    );

    class Program
    {
        static void Main()
        {
            var mutable1 = new Dictionary<string, int> { { "A", 0 }, { "B", 42 } };
            var mutable2 = new Dictionary<string, int> { { "A", 0 }, { "B", 42 } };

            var record1 = new MyRecord(mutable1, mutable1.ToImmutableDictionary());
            var record2 = new MyRecord(mutable2, mutable2.ToImmutableDictionary());

            Console.WriteLine(record1 == record2);
        }
    }
}
```

**Comment by diegofrata** [comment]
It's a very good point! Perhaps I can actually check if either the enumerables implement IDictionary<TKey, TValue> and avoid the extra allocation. Cheers for pointing that out!

**Comment by PatrikBak** [comment]
> 
> 
> It's a very good point! Perhaps I can actually check if both the enumerables implement IDictionary<TKey, TValue> and avoid the extra allocation. Cheers for pointing that out!

Agreed. You can also add

```CSharp
// If we have two collections with different counts, we can efficiently resolve their inequality
if (x is ICollection<T> xCollection && y is ICollection<T> yCollection && xCollection.Count != yCollection.Count)
    return false;

// If we have two sets, we can use its optimized SetEquals method
if (x is ISet<T> xSet && y is ISet<T> ySet)
    return xSet.SetEquals(ySet);
```

**Comment by diegofrata** [comment]
I implemented this and bumped the version to 1.0.0 since it's a breaking change and I think it's decent enough to be called 1.0!

Regarding the SetEquals, I did not use your suggestion as UnorderedEquality is not intended to have set semantics. That would require a new attribute, likely called SetEquality.

**Comment by diegofrata** [comment]
I actually went ahead and renamed the attributes to OrderedEquality and UnorderedEquality to make it more clear, ended up having to bump to 2.0.0 just to be consistent with SemVer.
